doc/rm/c_cpp_coding_style.md

```diff
@@ -173,6 +186,49 @@ This may change if we find cases where this initialization improves readability.
 When initializing an array, initializers *may* be designated when that makes the array more readable (e.g., lookup tables that are mostly zeroed). Mixing designated and undesignated initializers, or using nested initializers, is still
 forbidden.
 
+### Nonstandard Attributes
+
+The following nonstandard attributes may be used:
+*   `section(<name>)` to put a definition into a given object section.
+*   `weak` to make a symbol definition have weak linkage.
+*   `naked` to ensure a function has no prolog or epilog.
+*   `packed` to ensure a struct contains no padding.
+
+It is recommended that other nonstandard attributes are not used.
+
+All nonstandard attributes must be supported by both GCC and Clang.
+
+***The use of nonstandard attributes must be hidden by a macro definition starting `OT_`.***
+
+For instance, `__attribute__((section("foo")))` should not be used on a declaration.
+Instead, the attribute should be hidden behind a macro, like below.
+```c
+#define OT_SECTION(name) __attribute__((section(name)))
+//...
+void _crt(void) OT_SECTION(".crt");
+```
+
+This ensures changes to add compatibilty for other compilers are less invasive.
```

Comment:
What compilers other than GCC and clang do we expect? This macro seems to make code harder to understand for the vague possibility that we might want to support another compiler in the future.

Reply:
Very unlikely that we'll be using other compilers; that said, I support hiding `__attribute__`s behind macros when possible, because:
- Frankly, it's an ugly GNUism that makes code hard to read.
- It's good hygine.

Both of those are kind of subjective reasons, but ultimately I think style guide arbitration is pretty subjective... =P

Reply:
> What compilers other than GCC and clang do we expect? This macro seems to make code harder to understand for the vague possibility that we might want to support another compiler in the future.

I agree, I think `OT_SECTION` clouds the meaning.  It's not OT-specific but invites the reader to make that connection unless they're already familiar with the macros you're proposing.  Given the constraints that the attributes must already be GCC and Clang compatible I don't think putting them behind macros is valuable.

Reply:
Yeah, I'm swayed that it's easier to google for `__attribute__((section(".crt")))` than `OT_SECTION`. Yes it's a little uglier, but it's also unambiguous. We know that `__attribute__(…)` is supported by both GCC and Clang. 

I'm happy to revisit the issue when we want to support even more compilers, which are not compatible with the existing attribute syntax. I don't think we're at that point yet.